newrelic-agent/src/test/java/com/newrelic/agent/service/logging/LogSenderServiceImplTest.java

```diff
@@ -1,7 +1,247 @@
+        when(ServiceFactory.getServiceManager().getRPMServiceManager().getOrCreateRPMService(appName)).thenReturn(
+                analyticsData);
+
+        logSenderService.harvestHarvestables();
+
+        assertEquals(3, analyticsData.getEvents().size());
+
+        logSenderService.stop();
+
+        verify(txService, times(1)).removeTransactionListener(logSenderService.transactionListener);
+    }
+
+    @Test
+    public void testWithTransaction() throws Exception {
+        LogSenderServiceImpl logSenderService = createService(createConfig(null, 180));
+        Transaction transaction = Mockito.mock(Transaction.class);
+        when(ServiceFactory.getTransactionService().getTransaction(false)).thenReturn(transaction);
+
+        LogSenderServiceImpl.TransactionLogs logs = new LogSenderServiceImpl.TransactionLogs(
+                AgentConfigImpl.createAgentConfig(Collections.emptyMap()));
+        when(transaction.getLogEventData()).thenReturn(logs);
+        when(transaction.getApplicationName()).thenReturn(appName);
+        when(transaction.isInProgress()).thenReturn(true);
+
+        logSenderService.recordLogEvent(ImmutableMap.of("field", "value"));
+        logSenderService.recordLogEvent(ImmutableMap.of("field2", "value2"));
+        logSenderService.recordLogEvent(ImmutableMap.of("field3", "value3"));
+
+        MockRPMService analyticsData = new MockRPMService();
+        when(ServiceFactory.getServiceManager().getRPMServiceManager().getOrCreateRPMService(appName)).thenReturn(
+                analyticsData);
+
+        logSenderService.harvestHarvestables();
+
+        assertEquals(0, analyticsData.getEvents().size());
+        assertEquals(3, logs.events.size());
+    }
+
+    @Test
+    public void testTransactionHarvest() throws Exception {
+        LogSenderServiceImpl logSenderService = createService(createConfig(null, 180));
+        logSenderService.addHarvestableToService(appName);
+
+        Transaction transaction = Mockito.mock(Transaction.class);
+        when(ServiceFactory.getTransactionService().getTransaction(false)).thenReturn(transaction);
+
+        LogSenderServiceImpl.TransactionLogs logs = new LogSenderServiceImpl.TransactionLogs(
+                AgentConfigImpl.createAgentConfig(Collections.emptyMap()));
+        when(transaction.getLogEventData()).thenReturn(logs);
+        when(transaction.getApplicationName()).thenReturn(appName);
+        when(transaction.isInProgress()).thenReturn(true);
+
+        logSenderService.recordLogEvent(ImmutableMap.of("field", "value"));
+        logSenderService.recordLogEvent(ImmutableMap.of("field2", "value2"));
+        logSenderService.recordLogEvent(ImmutableMap.of("field3", "value3"));
+
+        // these should be filtered out
+        logSenderService.recordLogEvent(null);
+        logSenderService.recordLogEvent(Collections.emptyMap());
+
+        // these might get filtered out but don't cause the event to get filtered out
```

Comment:
These can't be created via Google ImmutableMap, but are still creatable and don't cause the even to be filtered out best I can tell.